python/lib/core/dmod/core/context.py

```diff
@@ -0,0 +1,607 @@
+from multiprocessing import RLock
+from multiprocessing import util
+from multiprocessing.context import BaseContext
+from traceback import format_exc
+
+from .decorators import version_range
+
+_PREPARATION_LOCK: RLock = RLock()
+
+SENTINEL = object()
+"""A basic sentinel value to serve as a true 'null' value"""
+
+T = typing.TypeVar("T")
+"""Some generic type of object"""
+
+Manager = typing.TypeVar("Manager", bound=managers.BaseManager, covariant=True)
+"""Any type of manager object"""
+
+ManagerType = typing.Type[Manager]
+"""The type of a manager object itself"""
+
+TypeOfRemoteObject = typing.Union[typing.Type[managers.BaseProxy], type]
+"""A wrapper object that is used to communicate to objects created by Managers"""
+
+_PROXY_TYPE_CACHE: typing.MutableMapping[typing.Tuple[str, typing.Tuple[str, ...]], TypeOfRemoteObject] = {}
+"""A simple mapping of recently created proxies to remote objects"""
+
+__ACCEPTABLE_DUNDERS = (
+    "__getitem__",
+    "__setitem__",
+    "__delitem__",
+    "__contains__",
+    "__call__",
+    "__iter__",
+    "__gt__",
+    "__ge__",
+    "__lt__",
+    "__le__",
+    "__eq__",
+    "__mul__",
+    "__truediv__",
+    "__floordiv__",
+    "__mod__",
+    "__sub__",
+    "__add__",
+    "__ne__",
+    "__get_property__",
+    "__set_property__",
+    "__del_property__"
+)
+"""A collection of dunder names that are valid names for functions on proxies to shared objects"""
+
+PROXY_SUFFIX: typing.Final[str] = "Proxy"
+"""
+Suffix for how proxies are to be named - naming proxies programmatically will ensure they are correctly referenced later
+"""
+
+
+@typing.runtime_checkable
+class ProxiableGetPropertyProtocol(typing.Protocol):
+    """
+    Outline for a class that can explicitly retrieve a property value
+    """
+    def __get_property__(self, key: str) -> typing.Any:
+        ...
+
+@typing.runtime_checkable
+class ProxiableSetPropertyProtocol(typing.Protocol):
+    """
+    Outline for a class that can explicitly set a property value
+    """
+    def __set_property__(self, key: str, value) -> None:
+        ...
+
+
+@typing.runtime_checkable
+class ProxiableDeletePropertyProtocol(typing.Protocol):
+    """
+    Outline for a class that can explicitly delete a property value
+    """
+    def __del_property__(self, key: str) -> None:
+        ...
+
+
+class ProxiablePropertyMixin(ProxiableGetPropertyProtocol, ProxiableSetPropertyProtocol, ProxiableDeletePropertyProtocol):
+    """
+    Mixin functions that allow property functions (fget, fset, fdel) to be called explicitly rather than implicitly
+    """
+    def __get_property__(self, key: str) -> typing.Any:
+        field = getattr(self.__class__, key)
+        if not isinstance(field, property):
+            raise TypeError(f"'{key}' is not a property of type '{self.__class__.__name__}'")
+
+        if field.fget is None:
+            raise Exception(f"Cannot retrieve the value for '{key}' on type '{self.__class__.__name__} - it is write-only")
```

Comment:
Never thought about this being possible. Not sure how useful it would be in practice, but still kind of interesting! Great edge case catch!